Migration of Shop parameters > Contact > Contacts page.

rokaszygmantas opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This PR migrates the contacts list page (Shop parameters > Contact > Contacts).
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/10622
| How to test?  | Compare to the legacy AdminContacts controller. Only the listing page is migrated, edit/add new form is not migrated by this PR. The page is hidden, but can be accessed directly via url: `/admin-dev/index.php/configure/shop/contacts/`

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/10777)
<!-- Reviewable:end -->

rokaszygmantas committed:
Add contacts controller with index action and route

rokaszygmantas committed:
Render contacts list, without javascript actions yet

rokaszygmantas committed:
Implement contacts grid javascript actions

rokaszygmantas committed:
Implement contact delete and bulk delete actions

rokaszygmantas committed:
Add demo mode restrictions

rokaszygmantas committed:
Update link generation for the new contact controller

rokaszygmantas committed:
Add a twig block around contacts grid

rokaszygmantas committed:
Fix back links from legacy contact form to link to migrated controller.
Forward the success/error/warning messages as well.

rokaszygmantas committed:
Rebuild assets

rokaszygmantas committed:
Fix cs

sarjon commented on classes/Link.php:
```diff
@@ -804,6 +804,7 @@ public function getAdminLink($controller, $withToken = true, $sfRouteParams = ar
                     'AdminMeta' => 'admin_meta',
                     // 'AdminWebservice' => 'admin_webservice', @todo: uncomment when grid and entity form are done.
                     'AdminBackup' => 'admin_backup',
+                    'AdminContacts' => 'admin_contacts',
```
not sure, but maybe it should be disabled until whole page is migrated? @matks wdyt?

matks replied:
Yes indeed. There are buttons on the "Contacts > Create/edit" page that redirect to the legacy controller and the behavior is very hard to change, so this page must remain hidden until the "Contacts > Create/edit" is migrated as well.

To allow the QA folks to validate, you must provide them the SF URL ;)

sarjon commented on classes/controller/AdminController.php:
```diff
@@ -4733,4 +4734,46 @@ private function getTabLinkFromSubTabs(array $subtabs)
+     * and redirect to the Symfony controller instead.
+     * Forwards the messages (if any) as well.
+     */
+    protected function redirectToSymfonyControllerWithMessages()
```
is it worth dealing with legacy part? the end goal is to remove it, so we can migrate CRUD in another Pr and only then enable new page maybe?

matks replied:
The goal is to remove the old controller link and manage retro-compatible legacy links using the method described in https://github.com/PrestaShop/PrestaShop/issues/10745

sarjon commented on src/Adapter/Support/ContactDeleter.php:
```diff
@@ -0,0 +1,77 @@
+/**
+ * Class ContactDeleter deletes contact records, using legacy code.
+ */
+final class ContactDeleter
```
it would be much better to implement it as a command, wdyt?

rokaszygmantas replied:
For sure, but I wanted this to be a grid-only PR. We can implement commands in a new PR once this is merged, to reduce the complexity

matks replied:
Nice approach 👍

sarjon commented on src/Core/Grid/Definition/Factory/ContactGridDefinitionFactory.php:
```diff
@@ -0,0 +1,227 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function getFilters()
```
i believe it should be `protected`?

rokaszygmantas replied:
yes, nice catch

sarjon commented on src/Core/Grid/Definition/Factory/ContactGridDefinitionFactory.php:
```diff
@@ -0,0 +1,227 @@
+                ])
+                ->setAssociatedColumn('actions')
+            )
+            ;
```
check indentation for `;`

sarjon commented on src/Core/Grid/Definition/Factory/ContactGridDefinitionFactory.php:
```diff
@@ -0,0 +1,227 @@
+                    'confirm_message' => $this->trans('Delete selected items?', [], 'Admin.Notifications.Warning'),
+                ])
+            )
+            ;
```
same indentation for `;`

sarjon commented on src/Core/Grid/Definition/Factory/ContactGridDefinitionFactory.php:
```diff
@@ -0,0 +1,227 @@
+                ->setName($this->trans('Export to SQL Manager', [], 'Admin.Actions'))
+                ->setIcon('storage')
+            )
+            ;
```
same indentation for ;

sarjon commented on src/PrestaShopBundle/Controller/Admin/Configure/ShopParameters/ContactsController.php:
```diff
@@ -0,0 +1,192 @@
+    /**
+     * Display the Contact creation form.
+     *
+     * @AdminSecurity("is_granted('create', request.get('_legacy_controller'))", message="You do not have permission to add this.")
```
if user has access to: create, update, delete. Does it provide access to the page? same for other actions.

rokaszygmantas replied:
If you don't have `read` access you cannot access the contact list page, as well as the contact creation form. Should it be accessible with such permissions?

matks replied:
I'm sorry I dont understand the issue here

sarjon commented on src/PrestaShopBundle/Resources/config/routing/admin/configure/shop_parameters/contacts.yml:
```diff
@@ -0,0 +1,46 @@
+        _controller: 'PrestaShopBundle:Admin\Configure\ShopParameters\Contacts:search'
+        _legacy_controller: AdminContacts
+
+
```
too many empty lines

sarjon commented on src/PrestaShopBundle/Resources/config/routing/admin/configure/shop_parameters/contacts.yml:
```diff
@@ -0,0 +1,46 @@
+        _legacy_controller: AdminContacts
+
+admin_contacts_edit:
+    path: /edit/{contactId}
```
```diff
-  path: /edit/{contactId}
+  path: /{contactId}/edit
```

sarjon commented on src/PrestaShopBundle/Resources/config/routing/admin/configure/shop_parameters/contacts.yml:
```diff
@@ -0,0 +1,46 @@
+        _legacy_controller: AdminContacts
+
+
+admin_contacts_create:
```
not sure if it should be `create` or `add`.

rokaszygmantas replied:
the controller action is called `createAction` as well so I used the same here for consistency. Do you think `add` should be used here?

sarjon replied:
i dont really know which one is more correct here :/ but in BO you click "Add new", not "Create new". Same with "Edit", you click "edit" and not "update", that's why you named it like `admin_contacts_edit` and not `admin_contacts_update`?

rokaszygmantas replied:
@matks @PierreRambaud maybe you have some insights about naming?

matks replied:
After giving it a thought, I think `create` is slightly better than `add` because:
- it is widely used in web applications
- it is used in CRUD acronym

So developers will understand quickly what a `create.html.twig` or a `createAction` is about.

sarjon replied:
@matks in that case, if we follow CRUD naming convention, we should rename `edit` to `update` as well?

matks replied:
I find `update` quite vague 😅  while `create` is really clear about what it does. I find `edit` clear too.

However this is my personal opinion, this is not a very good answer 😅

sarjon commented on src/PrestaShopBundle/Resources/config/routing/admin/configure/shop_parameters/contacts.yml:
```diff
@@ -0,0 +1,46 @@
+        contactId: \d+
+
+admin_contacts_delete:
+    path: /delete/{contactId}
```
```diff
- path: /delete/{contactId}
+  path: /{contactId}/delete
```

sarjon commented on src/PrestaShopBundle/Resources/config/routing/admin/configure/shop_parameters/contacts.yml:
```diff
@@ -0,0 +1,46 @@
+
+admin_contacts_delete:
+    path: /delete/{contactId}
+    methods: [GET]
```
better use POST

tomas862 replied:
or even better, DELETE ? :thinking:

rokaszygmantas replied:
I can change it to `DELETE` for single delete action, but bulk actions do not allow `DELETE` method so it would be inconsistent :no_mouth: 
https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Grid/Action/Bulk/Type/SubmitBulkAction.php#L60

matks replied:
@mickaelandrieu can we extend SubmitBulkAction to allow DELETE HTTP requests or is there a technical issue with this ?

sarjon reviewed (commented):
pretty small PR for a  whole page, nice. :+1:

tomas862 commented on src/Core/Grid/Query/ContactsQueryBuilder.php:
```diff
@@ -0,0 +1,133 @@
+            ->innerJoin('c', $this->dbPrefix . 'contact_lang', 'cl', 'c.id_contact = cl.id_contact')
+            ->innerJoin('c', $this->dbPrefix . 'contact_shop', 'cs', 'c.id_contact = cs.id_contact')
+            ->andWhere('cl.`id_lang`=' . $this->languageId)
+            ->andWhere('cs.`id_shop`=' . $this->shopId)
```
I'm pretty sure this will cause a problem in the "All shops" shop context. See https://github.com/PrestaShop/PrestaShop/pull/10774/files CurrencyQueryBuilder class for implementation which uses contextShopIds as array because in "All shops" there can be data displayed on the list from more then one shop.
Ofcourse I'm not sure it's the common case in all lists but it's worth to test the lists which has relations with shop entity. @sarjon wdyt?

rokaszygmantas replied:
fixed - thanks!

rokaszygmantas committed:
Minor fixes after review

rokaszygmantas committed:
Fix delete actions http method and contacts grid query

rokaszygmantas committed:
Fix translation domain

rokaszygmantas committed:
Fix query build in multishop context

rokaszygmantas committed:
Rebuild assets

matks commented:
Thank you @tomas862 and @sarjon for the reviews 😉

matks commented on src/Core/Grid/Query/ContactsQueryBuilder.php:
```diff
@@ -0,0 +1,138 @@
+    public function getCountQueryBuilder(SearchCriteriaInterface $searchCriteria)
+    {
+        $qb = $this->getQueryBuilder($searchCriteria->getFilters())
+            ->select('SQL_CALC_FOUND_ROWS')
```
I think `COUNT` is faster than `FOUND_ROWS`, is'nt it ?

PierreRambaud replied:
it is, and it will be easier to change if we decide to allow PgSQL, Oracle or anything else.

matks commented on src/Core/Grid/Query/ContactsQueryBuilder.php:
```diff
@@ -0,0 +1,138 @@
+                continue;
+            }
+
+            $qb->andWhere("$name LIKE :$name");
```
Although it is unlikely, I am not confortable with "$..." syntax for security reasons (SQL injections) :/
Can you protect this field using backquotes "`" ?

PierreRambaud replied:
using backquotes not completlty protect against sqli here. more if the `$name` contains a string like this: 
```
name` OR 1=1 OR `name
```

The secure way is to define each field you want to use, or check if these fields exist in ObjectModel

sarjon replied:
@PierreRambaud it would solve (possible) security issue, but add limitation. Dont forget that filters are dynamic, they can (and will) be extended/modified by modules, so checking on object model would not work. :/

PierreRambaud replied:
If someone wants to extend and modified this query, it will maybe not override or change this function but `getSearchQueryBuilder` or anything else that allows him to improve the query with `select()` or `join()` and not just adding `$name` that are not in the `select()`

sarjon replied:
ok, i think i see a possible improvement for it. But at the moment all filters (default & from modules) are handled here.

matks replied:
We found an issue to solve in 1.7.6 here 😃 ! "What is the proper way to handle search if a grid is extended ?"
Proper = easy, documented and reliable of course 😅

matks replied:
Issue created: https://github.com/PrestaShop/PrestaShop/issues/10823

sarjon replied:
i think that all we have to do is pass `SearchCriteria` to the hook so modules can handle their filters however they want to, wdyt?

matks replied:
We'll see in the issue (I have copied your suggestion there)

matks commented on src/PrestaShopBundle/Controller/Admin/Configure/ShopParameters/ContactsController.php:
```diff
@@ -0,0 +1,192 @@
+     *
+     * @AdminSecurity("is_granted('read', request.get('_legacy_controller'))")
+     *
+     * @Template("@PrestaShop/Admin/Configure/ShopParameters/Contact/Contacts/contacts.html.twig")
```
We just decided for naming conventions regarding controller actions, template names and route names & paths 😄 
https://github.com/PrestaShop/PrestaShop/issues/10597

Please follow them 😉

matks commented on src/PrestaShopBundle/Resources/config/routing/admin/configure/shop_parameters/contacts.yml:
```diff
@@ -0,0 +1,45 @@
+admin_contacts:
```
Please follow our new naming conventions ;) https://github.com/PrestaShop/PrestaShop/issues/10597

rokaszygmantas committed:
Hide the migrated page while it's unfinished

rokaszygmantas committed:
Fix delete method

rokaszygmantas committed:
Fix naming

rokaszygmantas committed:
Fix possible sql injection in grid query builder

rokaszygmantas committed:
Change grid count query

rokaszygmantas commented:
@matks thanks for comments. Fixed:
- hidden the page (removed from `Link::getAdminLink()`) since it's unfinished.
- fixed naming of controller actions, routes, templates according to new conventions.
- fixed filters sql to prevent possible sql injection.
- changed grid count query for better performance (changed `FOUND_ROWS()` to `COUNT()`).

matks commented:
Thansk @rokaszygmantas 😄 now we wait for QA validation

matks commented:
@rokaszygmantas I rebased your PR to allow merging as QA has validated. Once Travis build is green, I'll merge :)

EDIT: plus php-cs-fixer 😛

matks committed:
Apply php-cs-fixer

matks commented:
Thanks @rokaszygmantas !

rokaszygmantas commented:
thanks!
